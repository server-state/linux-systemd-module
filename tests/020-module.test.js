require('./_gen-shell-exec-mock');
const constants = require('../src/constants');

const shellExec = require('shell-exec');
const serverModule = require('../src');

function generateCommand(fullName, type, addProps) {
    return 'systemctl show ' + fullName +
        ' --property ' + constants.PROPERTIES.any +
        (constants.PROPERTIES[type] ? ',' + constants.PROPERTIES[type] : '') +
        (addProps ? ',' + addProps : '');
}


// ===== BEGIN TEST =====

describe('Test simple module behaviour', () => {
    it('should reject when no systemd units are given', () => {
        expect(serverModule()).rejects.toThrow();
    });

    it('should reject when a unit are given two or more times', () => {
        expect(serverModule([{ name: 'myService.service' }, { name: 'myService.service' }])).rejects.toThrow();
    });

    it('should reject with an empty unit name', () => {
        expect(serverModule([{ name: '' }])).rejects.toThrow();
    });
});

const INVALID_OPTIONS = [
    [
        'string',
        'I am a string'
    ],
    [
        'boolean',
        true
    ],
    [
        'number',
        3.14
    ],
    [
        'object',
        { prop1: 'prop1', prop2: 'prop2' }
    ],
    [
        'empty array',
        []
    ],
    [
        'invalid array elements',
        ['string', true, 3.14]
    ],
    [
        'empty unit name',
        [{name: null}]
    ],
    [
        'whitespace unit name',
        [{name: '  '}]
    ]
];

describe('Test invalid given options', () => {
    it.each(INVALID_OPTIONS)('should reject with invalid option format %s', (_type, value) => {
        expect(serverModule(value)).rejects.toThrow();
    });
});


const PROPERTY_DEFINITIONS = [
    [
        'service',
        'plain service unit',
        'plainService.service',
        'plainService.service',
        ''
    ],
    [
        'service',
        'implied service unit',
        'impliedService',
        'impliedService.service',
        ''
    ],
    [
        'service',
        'service unit with additional properties',
        'additionalService.service',
        'additionalService.service',
        'GuessMainPID'
    ],
    [
        'mount',
        'plain mount unit',
        'plainMount.mount',
        'plainMount.mount',
        ''
    ],
    [
        'mount',
        'mount unit with additional properties',
        'additionalMount.mount',
        'additionalMount.mount',
        'Delegate'
    ],
    [
        'slice',
        'plain slice unit',
        'plainSlice.slice',
        'plainSlice.slice',
        ''
    ],
    [
        'slice',
        'slice unit with additional properties',
        'additionalSlice.slice',
        'additionalSlice.slice',
        'ControlGroup'
    ],
    [ // virtual test type
        'other',
        'other unit type',
        'otherUnit.other',
        'otherUnit.other',
        ''
    ]
];

describe('Test given properties via argument calls', () => {
    it.each(PROPERTY_DEFINITIONS)('should call with %s properties while given %s', (type, _description, name, fullName, addProps) => {
        const argument = [{ name: name, addProps: addProps }];
        expect(serverModule(argument)).resolves.toMatchSnapshot();

        expect(shellExec).toBeCalledWith(generateCommand(fullName, type, addProps));
    });
});