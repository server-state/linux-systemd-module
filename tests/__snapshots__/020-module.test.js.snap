// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test given properties via argument calls should call with mount properties while given mount unit with additional properties 1`] = `
Object {
  "additionalMount.mount": Object {
    "ActiveEnterTimestamp": "",
    "ActiveEnterTimestampMonotonic": "",
    "ActiveState": "",
    "Delegate": "",
    "Description": "",
    "Id": "placeholderId",
    "LazyUnmount": false,
    "LoadState": "",
    "Result": "",
    "SubState": "",
    "Type": "",
    "UnitFileState": "",
    "UnitType": "mount",
    "What": "",
    "Where": "",
  },
}
`;

exports[`Test given properties via argument calls should call with mount properties while given plain mount unit 1`] = `
Object {
  "plainMount.mount": Object {
    "ActiveEnterTimestamp": "",
    "ActiveEnterTimestampMonotonic": "",
    "ActiveState": "",
    "Description": "",
    "Id": "placeholderId",
    "LazyUnmount": false,
    "LoadState": "",
    "Result": "",
    "SubState": "",
    "Type": "",
    "UnitFileState": "",
    "UnitType": "mount",
    "What": "",
    "Where": "",
  },
}
`;

exports[`Test given properties via argument calls should call with other properties while given other unit type 1`] = `
Object {
  "otherUnit.other": Object {
    "ActiveEnterTimestamp": "",
    "ActiveEnterTimestampMonotonic": "",
    "ActiveState": "",
    "Description": "",
    "Id": "placeholderId",
    "LoadState": "",
    "Result": "",
    "SubState": "",
    "UnitFileState": "",
    "UnitType": "other",
  },
}
`;

exports[`Test given properties via argument calls should call with service properties while given implied service unit 1`] = `
Object {
  "impliedService.service": Object {
    "ActiveEnterTimestamp": "",
    "ActiveEnterTimestampMonotonic": "",
    "ActiveState": "",
    "CleanResult": "",
    "Description": "",
    "Id": "placeholderId",
    "LoadState": "",
    "MainPID": 0,
    "MemoryCurrent": 0,
    "Nice": 0,
    "RemainAfterExit": false,
    "Restart": "",
    "Result": "",
    "SubState": "",
    "TasksCurrent": "",
    "Type": "",
    "UnitFileState": "",
    "UnitType": "service",
  },
}
`;

exports[`Test given properties via argument calls should call with service properties while given plain service unit 1`] = `
Object {
  "plainService.service": Object {
    "ActiveEnterTimestamp": "",
    "ActiveEnterTimestampMonotonic": "",
    "ActiveState": "",
    "CleanResult": "",
    "Description": "",
    "Id": "placeholderId",
    "LoadState": "",
    "MainPID": 0,
    "MemoryCurrent": 0,
    "Nice": 0,
    "RemainAfterExit": false,
    "Restart": "",
    "Result": "",
    "SubState": "",
    "TasksCurrent": "",
    "Type": "",
    "UnitFileState": "",
    "UnitType": "service",
  },
}
`;

exports[`Test given properties via argument calls should call with service properties while given service unit with additional properties 1`] = `
Object {
  "additionalService.service": Object {
    "ActiveEnterTimestamp": "",
    "ActiveEnterTimestampMonotonic": "",
    "ActiveState": "",
    "CleanResult": "",
    "Description": "",
    "GuessMainPID": "",
    "Id": "placeholderId",
    "LoadState": "",
    "MainPID": 0,
    "MemoryCurrent": 0,
    "Nice": 0,
    "RemainAfterExit": false,
    "Restart": "",
    "Result": "",
    "SubState": "",
    "TasksCurrent": "",
    "Type": "",
    "UnitFileState": "",
    "UnitType": "service",
  },
}
`;

exports[`Test given properties via argument calls should call with slice properties while given plain slice unit 1`] = `
Object {
  "plainSlice.slice": Object {
    "ActiveEnterTimestamp": "",
    "ActiveEnterTimestampMonotonic": "",
    "ActiveState": "",
    "Description": "",
    "Id": "placeholderId",
    "LoadState": "",
    "Result": "",
    "SubState": "",
    "TasksCurrent": "",
    "UnitFileState": "",
    "UnitType": "slice",
  },
}
`;

exports[`Test given properties via argument calls should call with slice properties while given slice unit with additional properties 1`] = `
Object {
  "additionalSlice.slice": Object {
    "ActiveEnterTimestamp": "",
    "ActiveEnterTimestampMonotonic": "",
    "ActiveState": "",
    "ControlGroup": "",
    "Description": "",
    "Id": "placeholderId",
    "LoadState": "",
    "Result": "",
    "SubState": "",
    "TasksCurrent": "",
    "UnitFileState": "",
    "UnitType": "slice",
  },
}
`;
